Spigot-Server-Patches/0675-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch

```diff
@@ -0,0 +1,362 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..a05b6344c2a3090c32de27415c274dd1dc4e820f
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/HopperPusher.java
+@@ -0,0 +1,88 @@
++package io.papermc.paper;
++
++import net.minecraft.server.AxisAlignedBB;
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.Blocks;
++import net.minecraft.server.Entity;
++import net.minecraft.server.MinecraftServer;
++import net.minecraft.server.TileEntity;
++import net.minecraft.server.TileEntityHopper;
++import net.minecraft.server.World;
++
++public interface HopperPusher {
++
++    default TileEntityHopper findHopper() {
++        BlockPosition pos = new BlockPosition(((Entity) this).locX(), ((Entity) this).locY(), ((Entity) this).locZ()).down();
++        TileEntityHopper hopper = getHopper(((Entity) this).getWorld(), pos);
++        if (hopper == null) {
++            return null;
++        }
++        AxisAlignedBB hopperBoundingBox = hopper.getHopperLookupBoundingBox();
++        if (hopperBoundingBox.intersects(((Entity) this)
++            .getBoundingBox())) {
++            return hopper;
++        }
++        return null;
++    }
++
++    boolean acceptItem(TileEntityHopper hopper);
++
++    default boolean tryPutInHopper() {
++        if (!((Entity) this).getWorld().paperConfig.isHopperPushBased) {
++            return false;
++        }
++        CheckData checkData = getCheckData();
++        Entity self = ((Entity)this);
++        if (self.locX() == self.lastX && self.locY() == self.lastY && self.locZ() == self.lastZ) { // samePositionAsLastTime
++            TileEntityHopper hopper = null;
++            if(checkData.getHopperPosition()!=null){
++                hopper = getHopper(((Entity) this).getWorld(), checkData.getHopperPosition());
++            } else if((net.minecraft.server.MinecraftServer.currentTick - checkData.getLastChecked()) > ((Entity) this).getWorld().paperConfig.entitySearchForHopperDelay){ // has waited long enough
++                hopper = findHopper();
++                if(hopper != null){
++                    checkData.setHopperPosition(hopper.getBlockPosition());
++                }
++                checkData.updateLastChecked();
++            }
++            return hopper != null && hopper.canAcceptItems() && acceptItem(hopper);
++        } else {
++            checkData.setHopperPosition(null);
++        }
++        return false;
++    }
++
++    static TileEntityHopper getHopper(World world, BlockPosition pos) {
++        if (world.getType(pos).getBlock() != Blocks.HOPPER) {
++            return null;
++        }
++        TileEntity tileEntity = world.getTileEntity(pos);
++        if (tileEntity instanceof TileEntityHopper) {
++            return (TileEntityHopper) tileEntity;
++        }
++        return null;
++    }
++
++    CheckData getCheckData();
++
++    class CheckData {
```

Comment:
Java 11 would be cool for protected classes  and getCheckData method for improved restriction, but it is as it is.